File: speedy_writeup/queue.ml
exception Empty;

type 'a queue = 'a list * 'a list
          
let push (q : 'a queue) (x : 'a) : 'a queue =
  match q with
  | [], back -> x::[], back
  | front, back -> front, x::back

let pop (q : 'a queue) : 'a queue * 'a = 
  match q with
  | [], _ -> raise Empty
  | x::[], back -> ((List.rev back, []), x)
  | x::xs, back -> ((xs, back), x)

File: speedy_writeup/queue.js
function Queue () {
    this.front = [];
    this.back = [];
    this.push = function (x) {
        if (this.front.length) {
            this.back.push(x);
        } else {
            this.front.push(x);
        }
    }
    this.pop = function () {
        if (this.front.length) {
            var ret = this.front.pop();
            if (! this.front.length) {
                this.front = this.back.reverse();
                this.back = [];
            }
            return ret;
        } else {
            this.front = this.back.reverse();
            this.back = [];
            return this.front.pop();
        }
    }
}
